File: resources/publication/pandoc/filters/tikzblock.lua
-- tikz-to-latex-advanced.lua
-- Pandoc filter to convert TikZ code blocks to raw LaTeX, removing surrounding document commands

function CodeBlock(block)
  -- Check if the block is TikZ code
  if block.classes[1] == 'tikz' then
    -- Define patterns to match \begin{document} and \end{document} with arbitrary spaces
    local beginDocPattern = "\\begin%s*{%s*document%s*}"
    local endDocPattern = "\\end%s*{%s*document%s*}"

    -- Remove everything from the beginning of the block to the end of the \begin{document} pattern
    local tikzCode = block.text:gsub(beginDocPattern .. ".*", "", 1)

    -- Remove everything from the \end{document} pattern to the end of the block
    tikzCode = tikzCode:gsub(".*" .. endDocPattern, "", 1)

    -- Construct the LaTeX code for the tikzpicture environment
    local latex = '\\begin{tikzpicture}[domain=0:4]\n' ..
                  tikzCode ..
                  '\n\\end{tikzpicture}'

    -- Return the block as raw LaTeX for Pandoc to process
    return pandoc.RawBlock('latex', latex)
  end
end

return {
  {CodeBlock = CodeBlock}
}

File: resources/publication/pandoc/filters/tikzblock.py
#! #! /usr/bin/env python
# -*- coding: utf-8 -*-

from pandocfilters import toJSONFilter, RawBlock
import sys
import re

def tikz_to_latex_advanced(key, value, format, meta):
    if key == 'CodeBlock':
        [[ident, classes, kvs], code] = value
        if 'tikz' in classes:
            begin_doc_pattern = r"\\begin\s*{\s*document\s*}"
            end_doc_pattern = r"\\end\s*{\s*document\s*}"
            tikz_code = re.sub(f"{begin_doc_pattern}.*", "", code, 1)
            tikz_code = re.sub(f".*{end_doc_pattern}", "", tikz_code, 1)
            # sys.stderr.write(tikz_code + "\n")
            latex = '\\begin{tikzpicture}[domain=0:4]\n' + tikz_code + '\n\\end{tikzpicture}'
            return RawBlock('latex', latex)

if __name__ == "__main__":
    toJSONFilter(tikz_to_latex_advanced)

